miden-lib/asm/sat/internal/epilogue.masm

```diff
@@ -21,12 +22,53 @@ const.CREATED_NOTE_HASHING_MEM_DIFF=1022
 #! - note_data_ptr is a pointer to the data section of the created note.
 #! - VAULT_HASH is the vault hash of the created note with index i.
 proc.compute_created_note_vault_hash
+    # insert note assets into the output vault
+    # ---------------------------------------------------------------------------------------------
     # duplicate note pointer - we will use this to save vault hash to memory
-    dup
-    # => [note_data_ptr, note_data_ptr]
+    dup dup exec.layout::get_created_note_num_assets
+    # => [num_assets, note_data_ptr, note_data_ptr]
 
     # get the number of assets from memory
-    dup exec.layout::get_created_note_num_assets
+    exec.layout::get_output_vault_root_ptr dup.2 exec.layout::get_created_note_asset_data_ptr dup 
+    # => [assets_start_ptr, assets_start_ptr, output_vault_root_ptr, num_assets, note_data_ptr, 
+    #     note_data_ptr]
+    
+    # compute end pointer 
+    dup.3 add swap
+    # => [assets_start_ptr, assets_end_ptr, output_vault_root_ptr, num_assets, note_data_ptr, 
+    #     note_data_ptr]
+
+    # loop over assets and insert them into input vault
+
+    # assess if we should loop
+    dup.1 dup.1 neq
+    # => [should_loop, assets_start_ptr, assets_end_ptr, output_vault_root_ptr, num_assets, 
+    #     note_data_ptr, note_data_ptr]
+
+    while.true
+        # duplicate output_vault_root_ptr
+        dup.2
+        # => [output_vault_root_ptr, assets_start_ptr, assets_end_ptr, output_vault_root_ptr, 
+        #     num_assets, note_data_ptr, note_data_ptr]
+
+        # read the asset from memory
+        padw dup.5 mem_loadw
+        # => [ASSET, output_vault_root_ptr, assets_start_ptr, assets_end_ptr, output_vault_root_ptr,
+        #     num_assets, note_data_ptr, note_data_ptr]
+
+        # insert asset into input vault
+        exec.account_vault::add_asset dropw
+        # => [assets_start_ptr, assets_end_ptr, output_vault_root_ptr, num_assets, note_data_ptr,
+        #     note_data_ptr]
+
+        # increment assets_start_ptr and asses if we should loop again
+        add.1 dup.1 dup.1 neq
+        # => [should_loop, assets_start_ptr, assets_end_ptr, output_vault_root_ptr, num_assets, 
+        #     note_data_ptr, note_data_ptr]
+    end
+
+    # clean stack
+    drop drop drop
```

Comment:
@bobbinth at the minute assets are pushed into the input vault in a separate loop (not the loop in which we generate the note vault hash).  This is because we would need to introduce a conditional block in the vault hash loop to check if the last asset is empty or not (to account for vault padding semantics).  I had thought that this additional conditional check would add more overhead than adding a separate loop.  Assessing the code I think this was the wrong decision in terms of code hygiene  and it would be best to have the conditional to check the last asset in the vault hashing loop.  I'm not too familiar with the program MAST hashing semantics to consider the implications of doing it inside the loop so I was hoping you could pitch advise.  I can always implement hashing inside the vault hash loop and compare benchmarks if that's preferable.

Reply:
I tried to implement this but I ran into some challenges and don't want to spend any more time on it when we have other more high priority items (mint + burn).  I propose we make an issue for this and come back to it.

Reply:
I think the current approach is fine. I am expecting the cost of building the output vault to be dominated by the cost of inserting items into the vault - so, the relative difference between the approach may not be all that big.

Maybe we should just add a `TODO` here so that we can come back here in the future and investigate if the other approach is meaningfully more efficient.

Reply:
Good point.  I will add a TODO.